### PR TITLE
[nieuwsblad] Add new extractor (Closes #7662)

### LINKED ISSUE #7662: Support for Het Nieuwsblad
* __URL:__ http://www.nieuwsblad.be/cnt/dmf20151124_01986463
* __Error:__ Unsupported URL: http://www.nieuwsblad.be/cnt/dmf20151124_01986463
* __Platform:__ openSUSE 13.1, x86_64, XFCE
* __Version:__ 2015.11.24
* __Video player:__ VRT Player (custom?)

```
sven@linux-etoq:/mnt/storage2/archive> ytdl -v http://www.nieuwsblad.be/cnt/dmf20151124_01986463
[debug] System config: []
[debug] User config: []
[debug] Command-line args: [u'--title', u'--continue', u'--retries', u'4', u'--write-info-json', u'--write-description', u'--write-thumbnail', u'--write-annotations', u'--all-subs', u'--ignore-errors', u'--merge-output-format', u'mkv', u'-f', u'bestvideo+bestaudio/best', u'-v', u'http://www.nieuwsblad.be/cnt/dmf20151124_01986463']
[debug] Encodings: locale ANSI_X3.4-1968, fs ANSI_X3.4-1968, out ANSI_X3.4-1968, pref ANSI_X3.4-1968
[debug] youtube-dl version 2015.11.24
[debug] Python version 2.7.6 - Linux-3.11.10-29-desktop-x86_64-with-SuSE-13.1-x86_64
[debug] exe versions: ffmpeg 2.8.2, ffprobe 2.8.2, rtmpdump 2.4
[debug] Proxy map: {}
[generic] dmf20151124_01986463: Requesting header
WARNING: Falling back on generic information extractor.
[generic] dmf20151124_01986463: Downloading webpage
[generic] dmf20151124_01986463: Extracting information
ERROR: Unsupported URL: http://www.nieuwsblad.be/cnt/dmf20151124_01986463
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/youtube_dl/extractor/generic.py", line 1282, in _real_extract
    doc = compat_etree_fromstring(webpage.encode('utf-8'))
  File "/usr/lib/python2.7/site-packages/youtube_dl/compat.py", line 248, in compat_etree_fromstring
    doc = _XML(text, parser=etree.XMLParser(target=etree.TreeBuilder(element_factory=_element_factory)))
  File "/usr/lib/python2.7/site-packages/youtube_dl/compat.py", line 237, in _XML
    parser.feed(text)
  File "/usr/lib64/python2.7/xml/etree/ElementTree.py", line 1642, in feed
    self._raiseerror(v)
  File "/usr/lib64/python2.7/xml/etree/ElementTree.py", line 1506, in _raiseerror
    raise err
ParseError: syntax error: line 3, column 0
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/youtube_dl/YoutubeDL.py", line 663, in extract_info
    ie_result = ie.extract(url)
  File "/usr/lib/python2.7/site-packages/youtube_dl/extractor/common.py", line 290, in extract
    return self._real_extract(url)
  File "/usr/lib/python2.7/site-packages/youtube_dl/extractor/generic.py", line 1887, in _real_extract
    raise UnsupportedError(url)
UnsupportedError: Unsupported URL: http://www.nieuwsblad.be/cnt/dmf20151124_01986463
```

### PR DESCRIPTION
**Pull request by fluxw42**
[nieuwsblad] Add new extractor (Closes #7662)

### CODE REVIEW

**Comment by yan12125** [comment]
Since #8497 landed, please move changes in ```youtube_dl/extractors/__init__.py``` to ```youtube_dl/extractors/extractors.py```. Check step 5 of the new [developer instructions](https://github.com/rg3/youtube-dl/blob/master/README.md#adding-support-for-a-new-site) for more information.

File: youtube_dl/extractor/nieuwsblad.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        webpage = self._download_webpage(url, video_id)
+
+        iframe_m = re.search(r'<script[^>]+src="(.+?kaltura.com.*?)"', webpage)
+        if iframe_m:
+            return self._extract_kaltura(url, webpage)
+
+        thumbnail = self._og_search_thumbnail(webpage)
+        title = self._og_search_title(webpage)
+
+        iframe_m = re.search(r'<iframe[^>]+src="(.+?vrt.be.*?)"', webpage)
```
Comment (yan12125) [inline youtube_dl/extractor/nieuwsblad.py]: Is this related to #11873?

File: youtube_dl/extractor/nieuwsblad.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        if iframe_m:
+            webpage = self._download_webpage(iframe_m.group(1), "vrt-iframe")
+            video_url = self._search_regex(r'sources.pdl = "(.*?)";', webpage, 'vrt-video')
+
+        iframe_m = re.search(r'<iframe[^>]+src="(.+?vmma.be.*?)"', webpage)
+        if iframe_m:
+            webpage = self._download_webpage(iframe_m.group(1), "vmma-iframe")
+            video_url = self._search_regex(r'<source src="(.*?)"', webpage, 'vmma-video')
```
Comment (yan12125) [inline youtube_dl/extractor/nieuwsblad.py]: I guess self._parse_html5_media_entries can be used here

File: youtube_dl/extractor/nieuwsblad.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        """ Delegate the video extraction to 'Kaltura' extractor """
+        kaltura_id = self._search_regex(r'entry_id\s*:\s*\"(.+?)\"', web_page, 'kaltura_id')
+        kaltura_wid = self._search_regex(r'wid\s*\:\s*\"(.+?)\"', web_page, 'kaltura_wid')
+        kaltura_uiconf_id = self._search_regex(r'uiconf_id\s*:\s*\"(.+?)\"', web_page, 'kaltura_uiconf_id')
+        kaltura_url = (
+            'https://cdnapisec.kaltura.com/index.php/kwidget/wid/{0}/uiconf_id/{1}/entry_id/{2}'
+                .format(kaltura_wid, kaltura_uiconf_id, kaltura_id)
+        )
+        url_with_source = smuggle_url(kaltura_url, {'source_url': url})
+        return self.url_result(url_with_source, 'Kaltura')
```
Comment (yan12125) [inline youtube_dl/extractor/nieuwsblad.py]: The pattern is quite similar to the generic extraction method ```KalturaIE._extract_url()```. I guess it can be used here.

**Comment by midas02** [comment]
@fluxw42 Yesterday, I noticed another use case on their website. Could you add it as well?
> '# Source: Vier.be
        {
            'url': 'http://www.nieuwsblad.be/cnt/dmf20170411_02829396',
            'md5': '3eb5c1ee0b0dcd428db9f97b9f1fb4a2',
            'info_dict': {
                'id': '02829396',
                'ext': 'mp4',
                'title': 'Dit is het nieuwste speeltje van Michel Van den Brande',
                'thumbnail': '????',
            }

Using your code:
> 
        iframe_m = re.search(r'<iframe[^>]+src="(.+?vier.be.*?)"', webpage)
        if iframe_m:
            webpage = self._download_webpage(iframe_m.group(1), "vier-iframe")
            video_url = self._search_regex(r'<source src="(.*?)"', webpage, 'vier-video')
the url becomes http://www.vier.be/video/partner/embed/v2/2658547/4b5a8c17b5358cb1d1b48e57966721bbef6df328/srnieuwsblad/asmh

Mapping can be found on http://www.vier.be/sites/all/modules/custom/sbs_video/js/player-bitdash.js
> videoFullPath = 'http://vod.streamcloud.be/' + videoApplication + '/mp4:_definst_/' + videoFilename + '.mp4/playlist.m3u8';

So it translates to http://vod.streamcloud.be/vier_vod/mp4:_definst_/theskyisthelimit/s4/20170411_sky_afl10_michel_suv_nieuw.mp4/playlist.m3u8 

But the thing is that vier.py doesn't have a parser for http://www.vier.be/video/partner/* links. So maybe it would be best to add the code for these links to vier.py, so the parsing can be deferred. What do you think @yan12125 ?

File: youtube_dl/extractor/nieuwsblad.py (hunk 2 above)
Comment (fluxw42) [inline youtube_dl/extractor/nieuwsblad.py]: I'm not really sure what you mean by 'codes' ... can you elaborate on this please?

**Comment by fluxw42** [comment]
@midas02 Just added support for embedded vier.be partner videos on nieuwsblad.be, using the `vier.py` extractor. Was it something like this you had in mind?

File: youtube_dl/extractor/nieuwsblad.py (hunk 2 above)
Comment (yan12125) [inline youtube_dl/extractor/nieuwsblad.py]: Here are some lines from medialaan.py:
```Python
        if not vod_id:
            player = self._parse_json(
                self._search_regex(
                    r'vmmaplayer\(({.+?})\);', webpage, 'vmma player',
                    default=''),
                video_id, transform_source=lambda s: '[%s]' % s, fatal=False)
            if player:
                video = player[-1]
                info = {
                    'id': video_id,
                    'url': video['videoUrl'],
                    'title': video['title'],
                    'thumbnail': video.get('imageUrl'),
                    'timestamp': int_or_none(video.get('createdDate')),
                    'duration': int_or_none(video.get('duration')),
                }
            else:
                info = self._parse_html5_media_entries(
                    url, webpage, video_id, m3u8_id='hls')[0]
                info.update({
                    'id': video_id,
                    'title': self._html_search_meta('description', webpage),
                    'duration': parse_duration(self._html_search_meta('duration', webpage)),
                })
```
There are also ```vmmaplayer()``` patterns in those vmma.be ```<iframes>```s, so I guess the logic in medialaan can be reused here.

**Comment by midas02** [comment]
fluxw42,
1. It occurred to me all Mediahuis newspapers use the same website technology: standaard, gva, and hbvl. It would make sense to have them in the same extractor.
I ran some tests, seems to work:
- http://www.gva.be/cnt/dmf20170412_02832220/blind-meisje-met-autisme-reageert-op-hartverwarmende-wijze-wanneer-ze-straatmuzikant-hoort-spelen
- http://www.hbvl.be/cnt/dmf20170413_02832693/video-bestuurder-krijgt-appelflauwte-en-rijdt-recht-woonkamer-binnen
- http://www.standaard.be/cnt/dmf20170413_02832378
I only encountered an issue with standaard.be. Their encoding of the kaltura url seems to be different (why???). Have a look, you'll notice that the Kaltura parser fails to identify the entry_id.
2. If you go for 1., it would make sense to rename the extractor Mediahuis, and add an IE_DESC = 'nieuwblad.be, standaard.be, ...'
3. These websites sometimes embed Youtube videos as well, but the generic extractor seems to be able to catch them. Not sure there is a need for encoding it separately.

Have a look. Thanks for your work so far. I'm thinking of writing one for hln/De Persgroep once you manage to finalise yours.

**Comment by fluxw42** [comment]
Hi, midas02. It makes a lot of sense to implement the Mediahuis sites under one umbrella extractor, especially since they use more or less the same way to embed their media. But I'm afraid I don't have any time to spare soon to implement this. Could you take this over for me?

**Comment by midas02** [comment]
Well, it doesn't require much work anymore. Just add these lines:
_VALID_URL = r'https?://(?:(?:www\.)?nieuwsblad\.be|(?:www\.)?gva\.be|(?:www\.)?standaard\.be|(?:www\.)?hbvl\.be)/.+?/dmf([0-9]+?)_(?P<id>[0-9]+)'
IE_DESC = 'Het Nieuwsblad, de Standaard, Gazet van Antwerpen, and Het Belang van Limburg'

Then there is the remark from @yan12125 on medialaan.py, but I don't understand what he's on about as well. You're just extracting the vmma-link and sending it to medialaan.py for extraction (not more than 3 lines!). You're not adding any code, so I don't understand what could be optimised there. Hopefully yan12125 can clarify.

Once the extractor goes in, I could take care of some maintenance if and when required.

**Comment by yan12125** [comment]
> You're just extracting the vmma-link and sending it to medialaan.py for extraction

That's indeed what I was looking for, but it's not how the current extractor is implemented.

File: youtube_dl/extractor/nieuwsblad.py (hunk 2 above)
Comment (midas02) [inline youtube_dl/extractor/nieuwsblad.py]: Parsing of vmma.be can be deferred to medialaan.py. This code works: https://github.com/rg3/youtube-dl/commit/67763af8595882485766bbb01c99897c14dd2c9e
@yan12125 The only thing I've never understood is the importance of video-ids for youtube-dl. Could you clarify, I can't find any documentation on it? In this example (www.nieuwsblad.be/cnt/dmf20151224_02036890) the referring page has an 'article id' of 02036890. The article embeds a vmma video using a 'vID' (https://syndication.medialaan.be/syndication/?vID=50534bb11fe5827eb25cad5c3445c5ae&sID=HNB&autoplay=false). But parsing that page, the json contains a video id "id":122781. So which one to use?
Please check https://github.com/rg3/youtube-dl/commit/67763af8595882485766bbb01c99897c14dd2c9e#R151, I don't know if this change is acceptable or not.

When it comes to mediahuis.py, the vmma link can then be deferred to mediahuis.py. Code (I added limburger.nl as well): https://github.com/rg3/youtube-dl/commit/9b34d3ede9e66c15f51468579db18a755e59809f
Yen, could you check https://github.com/rg3/youtube-dl/commit/9b34d3ede9e66c15f51468579db18a755e59809f#diff-fa9395c261329be01699b4ddff476e9cR168? Which is the preferred way of deferring the url, line 167 or line 168?

Comment (yan12125) [inline youtube_dl/extractor/nieuwsblad.py]: Video IDs are used in --download-archive. As long it's unique across videos belonging to the same extractor (i.e., different videos always map to different IDs), it's a good choice. If there are multiple satisfying IDs, it's up to you.

Comment (fluxw42) [inline youtube_dl/extractor/nieuwsblad.py]: For the mediahuis extractor I always used the `article id` from the user provided url, independent of the actual source of the video (kaltura, vmm, vrt, ...).

**Comment by fluxw42** [comment]
Whoops, I messed this one up ... fixing it in a minute.

**Comment by midas02** [comment]
@fluxw42 Could you remove the first test (Vmma - "Krijgt zieke Pauline (3) het mooiste kerstcadeau?") from the lot? These syndicated vmma links are always geo-restricted, and will make any testrun fail (I'm having to use a VPN to test it). Alternatively, there is a way of adding the expected result of the test ("HTTP Error 403: Forbidden"), but you'll have to look that one up.

**Comment by yan12125** [comment]
Any globally-available VMMA-based video? It's better to have one so that future fixes to this extractor won't accidentally break VMMA functionality.

**Comment by midas02** [comment]
Not to my knowledge. All syndicated links I've ever seen were all geo-restricted.

**Comment by fluxw42** [comment]
Anything I still have to do to get this merged?

**Comment by yan12125** [comment]
There are some commented codes:
```
# return self.url_result(video_url, 'Medialaan')
```
```
# 'id': str_or_none(video['id']) or video_id,
```
Any reason to keep them?

File: youtube_dl/extractor/medialaan.py [hunk 4]
```diff
@@ -17,11 +17,11 @@
...
     _VALID_URL = r'''(?x)
                     https?://
-                        (?:www\.)?
+                        (?:www\.|syndication\.)?
                         (?:
-                            (?P<site_id>vtm|q2|vtmkzoom)\.be/
+                            (?P<site_id>vtm|q2|vtmkzoom|vmma|medialaan)\.be/
                             (?:
-                                video(?:/[^/]+/id/|/?\?.*?\baid=)|
+                                video(?:/[^/]+/id/|/?\?.*?\baid=)|syndication/\?vID=|
```
Comment (yan12125) [inline youtube_dl/extractor/medialaan.py]: Better to have some URLs in _TESTS with ```'only_matching': True``` for changes to _VALID_URL

File: youtube_dl/extractor/mediahuis.py [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+    def _real_extract(self, url):
+        """ Extract the video info from the given mediahuis URL """
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+        thumbnail = self._og_search_thumbnail(webpage)
+        title = self._og_search_title(webpage)
+        description = self._og_search_description(webpage)
+        url_type = None
+
+        iframe_m = re.search(r'<script[^>]+src="(.+?kaltura.com.*?)"', webpage)
```
Comment (yan12125) [inline youtube_dl/extractor/mediahuis.py]: Better to use ```self._search_regex``` or ```self._html_search_regex``` instead of ```re.search```. The first two handles errors better.

File: youtube_dl/extractor/mediahuis.py [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+
+        iframe_m = re.search(r'<iframe[^>]+src="(.+?vier.be.*?)"', webpage)
+        if iframe_m:
+            video_url = (iframe_m.group(1))
+            url_type = 'url_transparent'
+
+        iframe_m = re.search(r'<iframe[^>]+src="(.+?vrt.be.*?)"', webpage)
+        if iframe_m:
+            webpage = self._download_webpage(iframe_m.group(1), "vrt-iframe")
+            video_url = self._search_regex(r'sources.pdl = "(.*?)";', webpage, 'vrt-video')
```
Comment (yan12125) [inline youtube_dl/extractor/mediahuis.py]: 1. Dots should be escaped
2. Better to use ```\s*``` or ```\s+``` instead of spaces in regular expressions for robustness

File: youtube_dl/extractor/mediahuis.py [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+
+        iframe_m = re.search(r'<iframe[^>]+src="(.+?vier\.be.*?)"', webpage)
+        if iframe_m:
+            video_url = (iframe_m.group(1))
+            url_type = 'url_transparent'
+
+        iframe_m = re.search(r'<iframe[^>]+src="(.+?vrt\.be.*?)"', webpage)
+        if iframe_m:
+            webpage = self._download_webpage(iframe_m.group(1), "vrt-iframe")
+            video_url = self._search_regex(r'sources.pdl\s*=\s*"(.*?)";', webpage, 'vrt-video')
```
Comment (yan12125) [inline youtube_dl/extractor/mediahuis.py]: Should be ```sources\.pdl``` here

File: youtube_dl/extractor/mediahuis.py [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+        {
+            'url': 'http://www.gva.be/cnt/dmf20170412_02832220/blind-meisje-met-autisme-reageert-op-hartverwarmende-wijze-wanneer-ze-straatmuzikant-hoort-spelen',
+            'md5': '54fc7fb24dd187adb25ccd698c45ddef',
+            'info_dict': {
+                'id': '02832220',
+                'ext': 'mp4',
+                'title': 'Blind meisje met autisme reageert op hartverwarmende wijze wanneer ze straatmuzikant hoort spelen',
+                'description': 'De zevenjarige Lacie is blind en lijdt aan autisme, maar haar liefde voor muziek is ze '
+                               'duidelijk niet verloren. Toen het meisje afgelopen vrijdag samen met...',
+                'thumbnail': 're:http.*jpg$',
```
Comment (yan12125) [inline youtube_dl/extractor/mediahuis.py]: Better to use ```r're:^https?://.*\.jpg$'``` here. Note that the ```r``` prefix is necessary or the code will fail on Python 3.6.

File: youtube_dl/extractor/mediahuis.py [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+
+        iframe_m = re.search(r'<iframe[^>]+src="(.+?vmma\.be.*?)"', webpage)
+        if iframe_m:
+            video_url = (iframe_m.group(1))
+            url_type = 'url_transparent'
+
+        # Source: flvpd.vtm.be/video.medialaancdn.be
+        iframe_m = re.search(r'<script.+?[^>]+videoUrl:\'(.+?)\'', webpage)
+        if iframe_m:
+            video_url = (iframe_m.group(1))
```
Comment (yan12125) [inline youtube_dl/extractor/mediahuis.py]: Any test case using this approach? I can't find it.

**Comment by yan12125** [comment]
Ah, are some Kaltura tests targetting the new codes in kaltura.py? If so those tests should be moved into kaltura.py.

File: youtube_dl/extractor/mediahuis.py (hunk 7 above)
Comment (fluxw42) [inline youtube_dl/extractor/mediahuis.py]: Fixed: Changed regex to match an actual ```.``` instead of any character.

File: youtube_dl/extractor/mediahuis.py (hunk 9 above)
Comment (fluxw42) [inline youtube_dl/extractor/mediahuis.py]: Included [test case with an url](http://www.gva.be/cnt/dmf20170105_02660060) for this particular pattern.

**Comment by fluxw42** [comment]
> Ah, are some Kaltura tests targetting the new codes in kaltura.py? If so those tests should be moved into kaltura.py.

Yes, they are included to cover the new code in kaltura.py, but they only use the static ```youtube_dl.extractor.kaltura.KalturaIE#_extract_url``` method. The actual url used to download the video has the same format as all kaltura urls (eg. ```kaltura:1467261:1_1z3nidfn```), so it wouldn't increase the test coverage if I add these as test cases.

Is there a way to include an automated test for the static ```_extract_url``` method?

**Comment by yan12125** [comment]
> The actual url used to download the video has the same format as all kaltura urls (eg. kaltura:1467261:1_1z3nidfn), so it wouldn't increase the test coverage if I add these as test cases.

I see. Then could you add some comments for what a test case is for? (with cdnapisec URL, for example)
